docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,216 @@
+
+The solution we proposed here is to start a storage layer from scratch. As Daniel Smith [mentioned](https://github.com/kubernetes/kubernetes/pull/28508#issuecomment-232508836)
+
+> Just to (probably unnecessarily) reiterate my position, I think we're going to need side-by-side implementations of both old and new storage styles, so we shouldn't end up locked into whatever we do now.
+
+We also believe we should start a side-by-side storage implementation to reach the goal we listed. That way, we won’t be affected by any existing issues, and it gives us the opportunities to rethink the design from the ground up, to achieve the goals more clearly. The only risk is that the effort will be completely wasted, if we cannot reach the goals.
+
+## Proposed Storage API
+
+```go
+
+// Storage stores and persists objects.
+// Storage provides read-write access to the objects based on string key.
+type Storage interface {
+	// Wait waits until the storage has seen any version >= given version.
+	// If version == 0, it provides a guarantee that once returning reads should see all writes from this interface.
+	Wait(ctx context.Context, version int64)
+
+	// Put puts an object with a key, and returns current object on given key.
+	// If conditions is not nil, it will do atomic Put. If compare failed, it will return existing object on given key.
+	// See "Conditions" docs for more details on compare logic and returned error.
+	Put(ctx context.Context, key StorageKey, obj runtime.Object, conditions *Conditions) (*KeyResponse, error)
+
+	// Delete deletes a key and its object, and returns deleted object on given key.
+	// conditions has the same guarantee as Put().
+	Delete(ctx context.Context, key StorageKey, conditions *Conditions) (*KeyResponse, error)
+
+	// Get gets the current object of given key.
+	// If version = 0, it will serve data from local cache.
+	// If version > 0, it will get the current state of the key at the time given version is committed.
+	// If no object exists on the key, it will return not found error.
+	Get(ctx context.Context, key StorageKey, version int64) (*KeyResponse, error)
+
+	// List lists all objects that has give prefix in keys and satisfies selectors.
+	// version has the same guarantee as Get().
+	List(ctx context.Context, prefix StorageKey, version int64, ss ...Selector) (*ListResponse, error)
+
+	// WatchPrefix watches a prefix after given version. If version is 0, we will watch from current state.
+	// It returns notifications of any keys that has given prefix.
+	// Given selectors, it returns events that contained object of interest, either of current and previous.
+	// If there is any problem establishing the watch channel, it will return error.
+	// After channel is established, any error will be returned from WatchChan immediately before it's closed.
+	WatchPrefix(ctx context.Context, prefix StorageKey, version int64, ss ...Selector) (WatchChan, error)
```

Comment:
I think I'd like a marker from "watch from the beginning" too.  If the value is not an integer (consistent with our API), then you'll need something besides `1`.